File: TabCommon/Library/src/main/java/com/nextzy/library/base/utils/permission/PermissionResult.java
package com.nextzy.library.base.utils.permission;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by TheKhaeng.
 */

public class PermissionResult{
    List<Permission> permissionList;

    public PermissionResult() {
    }

    public PermissionResult( List<Permission> permissionList) {
        this.permissionList = permissionList;
    }

    public boolean areAllPermissionsGranted() {
        if (isPermissionListAvailable()) {
            for (Permission permission : permissionList) {
                if (permission.isDenied()) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    public boolean isAnyPermissionPermanentlyDenied() {
        if (isPermissionListAvailable()) {
            for (Permission permission : permissionList) {
                if (permission.isDenied()) {
                    return true;
                }
            }
            return false;
        }
        return false;
    }

    private boolean isPermissionListAvailable() {
        return permissionList != null && permissionList.size() > 0;
    }

    public List<Permission> getDeniedPermissionResponses() {
        List<Permission> deniedPermissionList = new ArrayList<>();
        if (isPermissionListAvailable()) {
            for (Permission permission : permissionList) {
                if (permission.isDenied()) {
                    deniedPermissionList.add(permission);
                }
            }
        }
        return deniedPermissionList;
    }

    public List<Permission> getGrantedPermissionResponses() {
        List<Permission> grantedPermissionList = new ArrayList<>();
        if (isPermissionListAvailable()) {
            for (Permission permission : permissionList) {
                if (permission.isGranted()) {
                    grantedPermissionList.add(permission);
                }
            }
        }
        return grantedPermissionList;
    }

    public static class Permission {
        String permissionName;
        boolean isGranted;

        public Permission() {
        }

        public Permission(String permissionName, boolean isGranted) {
            this.permissionName = permissionName;
            this.isGranted = isGranted;
        }

        public String getPermissionName() {
            return permissionName;
        }

        public boolean isGranted() {
            return isGranted;
        }

        public boolean isDenied() {
            return !isGranted;
        }
    }
}

File: TabCommon/CommonLibrary/src/main/java/com/nextzy/library/base/utils/permission/PermissionResult.kt
package com.nextzy.library.base.utils.permission

import java.util.*

/**
 * Created by TheKhaeng.
 */

class PermissionResult(
        private var permissionList: List<Permission>) {

    private val isPermissionListAvailable: Boolean
        get() = permissionList.isNotEmpty()


    val isAnyPermissionPermanentlyDenied: Boolean
        get() {
            if (isPermissionListAvailable) {
                return permissionList.any { it.isDenied }
            }
            return false
        }


    val deniedPermissionResponses: List<Permission>
        get() {
            val deniedPermissionList = ArrayList<Permission>()
            if (isPermissionListAvailable) {
                permissionList.filterTo(deniedPermissionList) { it.isDenied }
            }
            return deniedPermissionList
        }

    val grantedPermissionResponses: List<Permission>
        get() {
            val grantedPermissionList = ArrayList<Permission>()
            if (isPermissionListAvailable) {
                permissionList.filterTo(grantedPermissionList) { it.isGranted }
            }
            return grantedPermissionList
        }


    fun areAllPermissionsGranted(): Boolean {
        if (isPermissionListAvailable) {
            return permissionList.none { it.isDenied }
        }
        return false
    }

    class Permission(permissionName: String,
                     isGranted: Boolean) {
        private var permissionName: String? = null
        var isGranted: Boolean = false

        val isDenied: Boolean
            get() = !isGranted
    }
}
